src/VoltageCalibration.cc

```diff
@@ -1111,6 +1139,50 @@ double * mattak::applyVoltageCalibration (int nSamples_wf, const int16_t * in, d
   return out;
 }
 
+double * mattak::applyVoltageCalibration(
+  int nSamples_wf, const int16_t * in, double * out, int start_window, bool isOldFirmware, int nResidPoints, const double * voltage_table, const std::array<std::vector<double>, 4096>* adc_table)
+
+{
+  if (!out) out = new double[nSamples_wf];
+
+  if (nSamples_wf % mattak::k::radiant_window_size)
+  {
+    std::cerr << "Not multiple of window size!" << std::endl;
+    return 0;
+  }
+
+  int nSamplesPerGroup = mattak::k::num_radiant_samples;
+  int nWindowsPerGroup = mattak::k::radiant_windows_per_buffer;
+  int sample_lab4, sample_A, sample_B;
+
+  sample_A = (start_window >= nWindowsPerGroup) * nSamplesPerGroup;
+
+  if (isOldFirmware)
+  {
+    nSamplesPerGroup /= 2;
+    nWindowsPerGroup /= 2;
+  }
+
+
+  for (int sample_wf = 0; sample_wf < nSamples_wf; sample_wf++)
+  {
+    sample_B = (sample_wf + start_window * mattak::k::radiant_window_size) % nSamplesPerGroup;
+
+    if (isOldFirmware)
+    {
+      sample_B += (sample_wf >= nSamplesPerGroup) * nSamplesPerGroup;
+    }
+
+    sample_lab4 = sample_A + sample_B;
+
+    // Get table for correct sample from pointer. Get pointer to point to address of first ele of table
+    const double * adc_table_sample = &adc_table->at(sample_lab4)[0];
```

Comment:
So I used a syntax with `[sample_lab4]` before. Maybe I had another bug in there but when I run it I would get a segmentation vault. I can have a look again but feel free to do it yourself too. On the other hand, maybe security at this point is more important than performance? In particular if we speed up the `adcToVolt`

Reply:
Well we can run it with at for a while and if it never crashes we can probably assume sample_lab4 is always in range :)

Reply:
but yes, I suspect the overhead here doesn't matter until at least adcToVolt is optimized

Reply:
> Well we can run it with at for a while and if it never crashes we can probably assume sample_lab4 is always in range :)

Puh, I am not sure if I would agree here.

Reply:
at( ) will crash if it's out of range (well, it will throw an exception, which we don't catch, or abort if exceptions are not disabled) so we'd find out pretty quickly!